sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -838,6 +839,7 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
              |$evPrim = $buffer.build();
            """.stripMargin
         }
+      case pudt: PythonUserDefinedType => castToStringCode(pudt.sqlType, ctx)
```

Comment:
This reminds me one thing: how shall we cast a UDT to string? Shall we just strip the UDT and cast the data as it's a catalyst value? or deserialize the catalyst value to customer object and call `toString`?

Note that `Dataset.show` should definitely call customer object's `toString`, but I think cast is different as it's an SQL operation.

Reply:
The second one (Deserializing the catalyst value to customer object and call toString). See: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L286

Reply:
As a note, I think `UserDefinedType` in Python is internal usage:

https://github.com/apache/spark/blob/3e40eb3f1ffac3d2f49459a801e3ce171ed34091/python/pyspark/sql/types.py#L642

I think we are fine as long as it shows more correct string form of defined UDTs within our projects.

Reply:
the java UDT is also internal, shall we make it consistent with python UDT?

Reply:
So, when casting any UDT to string, use `sqlType` casting.
And when `show()`ing Java UDT, use deserializing and the deserialized object's `toString()`, otherwise use casting to string. (I guess we shouldn't use the deserialized object's `toString()` for Python-only UDT which shows the raw string I mentioned in the description.)
Is that right?

Reply:
New thought: since UDT is not finalized yet(internal only), the only thing we care about is to have a reasonable string representation.

It's unclear that UDT class always have a reasonable `toString`, and `UDT.deserialize` may be pretty slow, how about we always use `UDT.sqlType` to string casting and `showString`?

Reply:
But, `VectorUDT.sqlType` has non-array formats:
https://github.com/apache/spark/blob/1c76a91e5fae11dcb66c453889e587b48039fdc9/mllib/src/main/scala/org/apache/spark/ml/linalg/VectorUDT.scala#L88

In this case, how do we convert `VectorUDT` data into array-lie strings (e.g., [0, 1, 2, ...])?

Reply:
Like the python UDT, we cursively call `castToStringCode(pudt.sqlType, ctx)`, does it work?

Reply:
Yes, it works to cast to string.

Btw, as for `VectorUDT`, seems like `DenseVector` and `SparseVector` override `toString()` at least for `show()` on purpose(?):
https://github.com/apache/spark/blob/74c17353bb6372b123c5aee1b6d58a21de36f99a/python/pyspark/ml/classification.py#L1497-L1503

If we also use cast to string for `show()`, the result will be like:

```
+-----------------+----------+
|         features|prediction|
+-----------------+----------+
|[1,,, [1.0, 0.0]]|       1.0|
|[1,,, [0.0, 0.0]]|       0.0|
+-----------------+----------+
```

I'm not sure we can change the string here.

Reply:
You suggested like this? https://github.com/apache/spark/compare/master...maropu:SPARK-23054-2

If so, this just dumps an internal structure;
```
scala> val df1 = Seq((1, Vectors.dense(Array(1.0, 2.0, 3.0)))).toDF("a", "b")
scala> df1.selectExpr("CAST(b AS STRING)").show(false)
+----------------------+
|b                     |
+----------------------+
|[1,,, [1.0, 2.0, 3.0]]|
+----------------------+

scala> val df2 = Seq((1, Vectors.sparse(3, Array(0, 2), Array(1.0, 3.0)))).toDF("a", "b")
scala> df2.selectExpr("CAST(b AS STRING)").show(false)
+--------------------------+
|b                         |
+--------------------------+
|[0, 3, [0, 2], [1.0, 3.0]]|
+--------------------------+
```

Reply:
maybe we should do the same thing for python UDT in the future, and leave it for now.

Reply:
How about what I suggested at https://github.com/apache/spark/pull/20306#discussion_r162269190?

Reply:
it looks like `[1,,, [1.0, 2.0, 3.0]]` is not a good string representation, and we should assume UDT defines a reasonable `toString`.

Reply:
I see, sounds good. Thanks.